### PR TITLE
clearer resolution of variables + add warning

### PR DESCRIPTION
**Pull request by bgregoir**
this PR solve #604.

### CODE REVIEW

**Comment by bgregoir** [comment]
Any comment ?

**Comment by vbgl** [comment]
The refactoring looks OK. I don’t understand the new warnings though. IMHO they do not belong to pre-typing but to your IDE or the so-called linter.

**Comment by bgregoir** [comment]
This is true, but we spend two hours (with Tiago) to debug a program due to this absence of warning.
Another problem, for used declaration, is that we do not keep track of the local variables declaration, this mean that it is hard to write a linter that will warn stuff after pretyping (since the info does not exists anymore).

**Comment by bgregoir** [comment]
Maybe I can remove the warning from that file and store the needed info somewhere else. 
Then I can add a linter.ml that will for the moment only print that warning.
But we can extent it to print more.
Should I do that ?

File: compiler/src/pretyping.ml [hunk 1]
```diff
@@ -369,30 +371,51 @@ end = struct
...
+      { env with e_decls = P.MIparam d :: env.e_decls } 
+    
+    let push_local (env : 'asm env) (v : P.pvar) =
+      env.e_declared := P.Spv.add v !(env.e_declared);
+      push_core env v Slocal
 
-    let find (x : A.symbol) (env : 'asm env) =
-      Map.Exceptionless.find x env.e_vars
-  end
+    let push_implicit (env : 'asm env) (v : P.pvar) =
```
Comment (vbgl) [inline compiler/src/pretyping.ml:408]:
```suggestion
    let push_implicit (env : 'asm env) (v : P.pvar) =
```

**Review by vbgl — APPROVED** [review]
Please add a CHANGELOG entry.

**Pull request merged**